src/common/dao/immutable.go

```diff
@@ -0,0 +1,94 @@
+package dao
+
+import (
+	"fmt"
+
+	"github.com/astaxie/beego/orm"
+	"github.com/goharbor/harbor/src/common"
+	"github.com/goharbor/harbor/src/common/models"
+	"strings"
+)
+
+// CreateImmutableRule creates the Immutable Rule
+func CreateImmutableRule(ir *models.ImmutableRule) (int64, error) {
+	ir.Enabled = true
+	ir.Decoration = getValidDecoration(ir.Decoration)
+	if exceedMaxImmutableRuleCount(ir.ProjectID) {
+		return 0, fmt.Errorf("immutable tag rule count exceed %v", common.MaxRuleCount)
+	}
+	o := GetOrmer()
+	return o.Insert(ir)
+}
+
+func getValidDecoration(decoration string) string {
+	if strings.EqualFold(decoration, common.DecorationInclude) || strings.EqualFold(decoration, common.DecorationExclude) {
+		return strings.ToLower(decoration)
+	}
+	return common.DecorationInclude
+}
+
+func exceedMaxImmutableRuleCount(projectID int64) bool {
+	o := GetOrmer()
+	var count int
+	o.Raw("select  count(*) cnt from immutable_tag_rule where project_id = ? ", projectID).QueryRow(&count)
+	return count >= common.MaxRuleCount
+}
+
+// UpdateImmutableRule update the immutable rules
+func UpdateImmutableRule(projectID int, ir *models.ImmutableRule) (int64, error) {
+	o := GetOrmer()
+	ir.Decoration = getValidDecoration(ir.Decoration)
+	return o.Update(ir, "RepoFilter", "TagFilter", "Decoration")
+}
+
+// EnableImmutableRule enable/disable immutable rules
+func EnableImmutableRule(id int64, enabled bool) (int64, error) {
```

Comment:
the name is confusing? If I want to disable any of rule, I still have to call EnableImmutableRule()

Reply:
change to ToggleImmutableRule